### PR TITLE
WIP: loose parameter as dict

### PR DESCRIPTION
**Pull request by joewalter**
WIP addressing #2163 and #4308. 

Added a new private function, which checks the input to loose and adapts it if necessary (and also converts forward to surface oriented if required). Currently, only loose = None and loose=1. are supported for vol source spaces. Please let me know if loose orientation constraint should also be available for vol source spaces.

### CODE REVIEW

File: mne/forward/forward.py [hunk 1]
```diff
@@ -875,6 +875,70 @@ def write_forward_meas_info(fid, info):
     end_block(fid, FIFF.FIFFB_MNE_PARENT_MEAS_FILE)
 
 
+def _check_loose(forward, loose):
+    """Check loose parameter input."""
+    if is_fixed_orient(forward):
```
Comment (agramfort) [inline mne/forward/forward.py]: does it mean I always have a warning when working with fixed ori?

I think we should not have a warning when user set loose to None explicitly.

Comment (joewalter) [inline mne/forward/forward.py]: I added `and loose is not None`, so the warning will now only appear if loose is not None

File: mne/forward/forward.py [hunk 2]
```diff
@@ -875,6 +875,70 @@ def write_forward_meas_info(fid, info):
     end_block(fid, FIFF.FIFFB_MNE_PARENT_MEAS_FILE)
 
 
+def _check_loose(forward, loose):
+    """Check loose parameter input."""
+    if is_fixed_orient(forward) and loose is not None:
+        warn('Ignoring loose parameter with forward operator '
+             'with fixed orientation.')
```
Comment (agramfort) [inline mne/forward/forward.py]: Say to set it to None to avoid this warning

File: mne/forward/forward.py [hunk 3]
```diff
@@ -875,6 +875,70 @@ def write_forward_meas_info(fid, info):
...
+        loose = None
+
+    if loose is not None:
+        if isinstance(loose, float):
+            if not (0 <= loose <= 1):
+                raise ValueError('Loose value should be smaller than 1 and '
+                                 'bigger than 0, or None for not loose '
+                                 'orientations.')
```
Comment (agramfort) [inline mne/forward/forward.py]: say what you got

File: mne/forward/forward.py [hunk 4]
```diff
@@ -875,6 +875,70 @@ def write_forward_meas_info(fid, info):
...
+            if loose < 1 and not forward['surf_ori']:
```
Comment (agramfort) [inline mne/forward/forward.py]: I would say

if loose is not None and not forward['surf_ori']:

File: mne/forward/forward.py [hunk 5]
```diff
@@ -894,29 +961,25 @@ def compute_orient_prior(forward, loose=0.2, verbose=None):
...
-
+    forward, loose = _check_loose(forward, loose)
     orient_prior = np.ones(n_sources, dtype=np.float)
-    if (not is_fixed_ori) and (loose is not None) and (loose < 1):
-        logger.info('Applying loose dipole orientations. Loose value '
-                    'of %s.' % loose)
-        orient_prior[np.mod(np.arange(n_sources), 3) != 2] *= loose
-
+    if ((loose is not None) and
+            not all([item in [1., None] for item in loose.values()])):
```
Comment (agramfort) [inline mne/forward/forward.py]: I think we should force loose=None rather than accepting 1.

File: mne/minimum_norm/tests/test_inverse.py [hunk 6]
```diff
@@ -367,9 +367,9 @@ def test_make_inverse_operator_free():
     assert_raises(ValueError, make_inverse_operator, evoked.info, fwd_2,
                   noise_cov, depth=None)
 
-    # for free ori inv, loose=None and loose=1 should be equivalent
+    # for free ori inv, loose=None and loose=1. should be equivalent
```
Comment (agramfort) [inline mne/minimum_norm/tests/test_inverse.py]: @Eric89GXL  are you ok to force loose=None for non surf ori forward?

**Comment by Eric89GXL** [comment]
Why not just always use float? It seems cleaner to me.

**Comment by agramfort** [comment]
None is more explicit to say "ignore this param"

**Comment by Eric89GXL** [comment]
Maybe from an internal coding perspective. But it seems less explicit to
the user in terms of what the effective value / outcome is, and less simple
in terms of values accepted.

To put it differently, in the future if we have a parameter in another
function that varies smoothly between two extremes, where one extreme
effectively does nothing, should we always disallow that extreme value and
force people to use None?

**Comment by Eric89GXL** [comment]
... For example, any regularization parameter.

**Comment by Eric89GXL** [comment]
The one way I can see it making sense is if for free orientation forward
None meant "don't modify", and for fixed orientation None also meant "don't
modify", but in that case we shouldn't disallow the equivalent numerical
value (0 or 1) in either case.

**Comment by agramfort** [comment]
>
> we shouldn't disallow the equivalent numerical value (0 or 1) in either
> case.
>

ok for me but I would document / recommend the use of None.

**Comment by joewalter** [comment]
Do we want to allow a loose orientation constraint for discrete source spaces? IIRC they are used for defining volume source spaces, so I am wondering whether we should treat them similarly to volume source spaces?

**Comment by agramfort** [comment]
I would say no. Loose ori only makes sense for surface oriented source spaces

Alex

**Comment by Eric89GXL** [comment]
I would say yes. Discrete shouldn't be used for volumes, volume type should
be used for that. IIUC discrete is used for anything custom the user wants,
and normals can easily be properly defined there. In fact the only two
things the user provides for Discrete type are locations and normals.

**Comment by agramfort** [comment]
ok for me then as long as volume does not use discrete source space

**Comment by Eric89GXL** [comment]
Argh, we do have these lines:
```
    if mri is not None:
        for s in sp:
            _add_interpolator(s, mri, add_interpolator)
    elif sp[0]['type'] == 'vol':
        # If there is no interpolator, it's actually a discrete source space
        sp[0]['type'] = 'discrete'
```
So we change `vol` type to `discrete` if there is no `mri`. This seems a bit odd to me, since it's still defined on a volume grid in volume space. But maybe because it's because you can't easily / properly interpolate into a volume for viz...?

I think this use case is likely rare (?), but the case of user-defined discrete source spaces might also be rare. To be safe, I'm okay with forcing free ori for `discrete` for now -- we can always add support for it later if we find users need it.

**Comment by joewalter** [comment]
Ok, I'll force free orientation with my next commit.

**Comment by joewalter** [comment]
With regard to sparse solvers I thought of normalizing the vector (weight_tan1, weight_tan2, weight_normal) in order not to bias the solution towards sources with free orientation (I'd add a sparse_inv flag, so that nothing changes for MNE-type methods)

**Comment by Eric89GXL** [comment]
> I'd add a sparse_inv flag, so that nothing changes for MNE-type methods

Why wouldn't you also want this for MNE methods, at least as an option? That way if I have 1000 fixed-orientation surface source-space points and 1000 free-orientation volume points, you could achieve roughly equivalent activation mapping into both (spatial distribution considerations aside), right?

**Comment by agramfort** [comment]
same concern as Eric. It seems like a bug
to work differently with L2 or sparse solvers.

**Comment by joewalter** [comment]
When we agree to only support loose or fixed orientation for surface-based source spaces, we can make the handling of loose much simpler, as we don't need the dict (so the check_loose function is much simpler) and we just add an info that loose parameter is ignored for volume and discrete source spaces.

**Comment by agramfort** [comment]
we could do this for now until there is a motivation to have loose for
discrete source space.
let's keep it simple

**Comment by Eric89GXL** [comment]
Sounds like a plan.

Does this also take care of #4313?

Is it possible/easy to update some example? Even a mixed-source-space MNE example might benefit from it.

File: mne/forward/forward.py (hunk 3 above)
Comment (agramfort) [inline mne/forward/forward.py]: @joewalter reminder here

File: mne/forward/forward.py [hunk 7]
```diff
@@ -875,6 +889,33 @@ def write_forward_meas_info(fid, info):
...
+                                 'bigger than 0, or None for not loose '
+                                 'orientations.')
+            if loose < 1 and not forward['surf_ori']:
+                warn('Forward operator is not oriented in surface '
+                     'coordinates. A loose inverse operator requires a '
+                     'surface-oriented forward operator with free '
+                     'orientation.')
+                forward = convert_forward_solution(forward, surf_ori=True)
+        else:
+            raise ValueError('loose value must be None for not loose '
```
Comment (agramfort) [inline mne/forward/forward.py]: loose must be None for not loose orientations ?

this is unclear what you mean here.

File: mne/forward/forward.py [hunk 8]
```diff
@@ -894,29 +939,24 @@ def compute_orient_prior(forward, loose=0.2, verbose=None):
...
-        orient_prior[np.mod(np.arange(n_sources), 3) != 2] *= loose
-
+    if loose is not None and loose < 1.:
+        if loose == 0.0:
+            logger.info('Applying fixed dipole orientations for surface-based '
+                        'source spaces.')
+        else:
+            logger.info('Applying loose dipole orientations for surface-based '
+                        'source spaces with loose=%f.' % loose)
+        idx_start = 0
```
Comment (agramfort) [inline mne/forward/forward.py]: can you add here
```
assert not is_fixed_orient(forward)
```
to avoid futur bugs

**Comment by agramfort** [comment]
besides so far so good

@joewalter can you see why travis is not happy?

**Comment by joewalter** [comment]
added for VolumeSourceEstimate and plotting and save methods for MixedSourceEstimate

Known issues:
- For `MixedSourceEstimate`, MNE assumes that first 2 source spaces are of type "surf", but there is no function, which checks this order
- `_make_stc` needs `src` as an input parameter. The current function give wrong output if the source space consists of exacly 2 source spaces of type "vol" (alternatively we could combine all volume source spaces into a single volume)

**Comment by joewalter** [comment]
#4372 is currently a blocker for this PR, as the tests fail on Travis due to the way fixed forwards and handled and generated. So, solving #4372 should also improve this PR.

**Comment by joewalter** [comment]
Any good ideas how to solve the issue that `_make_stc` requires info on `src` as an input parameter?
- Giving `src` as an input parameter to `_make_stc` is an API change, but at least it's a private function
- Alternatively, we could group the vertnos corresponding to the two surface source spaces in a separate list such that vertno is [[left_vertno, right_vertno], vol1_vertno, vol2_vertno, ..., volN_vertno], but this is also an API change

**Comment by Eric89GXL** [comment]
> Giving src as an input parameter to _make_stc is an API change, but at least it's a private function

I don't see any problem with this, assuming that the functions that use `_make_stc` already have `src` as a param.

**Comment by joewalter** [comment]
Unfortunately, there's one place where this doesn't work (_BaseSourceEstimate.bin).

Is it allowed to add the type as a new key to _BaseSourceEstimate with default `None` and which is then set to SourceEstimate, VolSourceEstimate, or MixedSourceEstimate. While writing an stc to disk, the type would only determine the filename, so there shouldn't be problems with compatibility.

**Comment by Eric89GXL** [comment]
I'm not quite sure what you mean. But what might work is having each subclass call the super class's (`_BaseSourceEstimate`'s) method with some additional argument that is never exposed to the user. Is that what you mean?

**Comment by larsoner** [comment]
@joewalter what is left here? Just a rebase and a few minor remaining comments?

**Comment by larsoner** [comment]
I don't think we should do this until 0.16 so we have some time to finish and then test, pushing to 0.16

**Comment by joewalter** [comment]
I think it is a good idea to push this to 0.16 so we have enough time for improvements and tests

**Comment by agramfort** [comment]
let's not close. We'll take some code from here.

**Comment by joewalter** [comment]
I think to make this one easier, we could split it in several PR. So we can faster merge parts of it, making rebasing simpler

**Comment by agramfort** [comment]
yes agreed

**Comment by larsoner** [comment]
Perhaps the first split could take care of #4604. That would be nice to have for 0.15

**Comment by larsoner** [comment]
Removing 0.16 tag

**Comment by larsoner** [comment]
This PR is way out of date at this point. Now that almost all code is refactored to use the same `loose` and `depth` paths, we should probably start with a fresh PR for this.

**Pull request closed without merging**